src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,297 @@
+          AsyncContextFrame::attachContext(js, promise);
+        }
+        break;
+      }
+      case v8::PromiseHookType::kBefore: {
+        // The kBefore event is triggered immediately before a Promise continuation.
+        // We use it here to enter the AsyncContextFrame that was associated with the
+        // promise when it was created.
+        KJ_IF_MAYBE(frame, AsyncContextFrame::tryGetContext(js, promise)) {
+          isolateBase.pushAsyncFrame(*frame);
+        } else {
+          // If the promise does not have a frame attached, we assume the root
+          // frame is used. Just to keep bookkeeping easier, we still go ahead
+          // and push the frame onto the stack again so we can just unconditionally
+          // pop it off in the kAfter without performing additional checks.
+          isolateBase.pushAsyncFrame(*isolateBase.rootAsyncFrame);
+        }
+        // We do not use AsyncContextFrame::Scope here because we do not exit the frame
+        // until the kAfter event fires.
+        break;
+      }
+      case v8::PromiseHookType::kAfter: {
+  #ifdef KJ_DEBUG
+        KJ_IF_MAYBE(frame, AsyncContextFrame::tryGetContext(js, promise)) {
+          // The frame associated with the promise must be the current frame.
+          KJ_ASSERT(frame == &currentFrame);
+        } else {
+          KJ_ASSERT(currentFrame.isRoot(js));
+        }
+  #endif
+        isolateBase.popAsyncFrame();
+
+        // If the promise has been rejected here, we have to maintain the association of the
+        // async context to the promise so that the context can be propagated to the unhandled
+        // rejection handler. However, if the promise has been fulfilled, we do not expect
+        // the context to be used any longer so we can break the context association here and
+        // allow the opaque wrapper to be garbage collected.
+        if (!isRejected()) {
+          auto handle = js.getPrivateSymbolFor("asyncResource"_kj);
+          check(promise->DeletePrivate(js.v8Isolate->GetCurrentContext(), handle));
+        }
+
+        break;
+      }
+      case v8::PromiseHookType::kResolve: {
+        // This case is a bit different. As an optimization, it appears that v8 will skip
+        // the kInit, kBefore, and kAfter events for Promises that are immediately resolved (e.g.
+        // Promise.resolve, and Promise.reject) and instead will emit the kResolve event first.
+        // When this event occurs, and the promise is rejected, we need to check to see if the
+        // promise is already wrapped, and if it is not, do so.
+        if (!currentFrame.isRoot(js) && isRejected() &&
+            AsyncContextFrame::tryGetContext(js, promise) == nullptr) {
+          AsyncContextFrame::attachContext(js, promise);
+        }
+        break;
+      }
+    }
+  }, [isolate](Value&& exception) {
+    isolate->ThrowException(exception.getHandle(isolate));
```

Comment:
Another reason we should use something other than `js.tryCatch()` is because we don't want to catch and then rethrow the exception, we'd rather it just flow through. So we sort of want to just `catch (JsExceptionThrown&) { return; }` -- catching `JsExceptionThrown` implies that the exception is already scheduled on the isolate so we don't need to do it again. OTOH if we catch `kj::Exception` we might have to do something with that.

Meanwhile, though... where does this exception even go? There isn't really a caller to propagate it to. Does V8 eat it? Does it abort the process itself? We should probably test what actually happens by temporarily adding some code to throw an exception...

Reply:
V8 bubbles the exception up. For instance, if I make it so that the promiseHook unconditionally just throws an error every time, the test reports an unhandled exception during startup.

```
1m10	2023.1.0-28-g17d91a0f9-dev error   workerd/io/worker.c++:382 script startup threw exception; id = simple-als-works; description = jsg.Error: boom; trace = ; ownerId = 1212; zoneId = test-zone; scriptId = simple-als-works; cordon = paid; process = shared
```